### PR TITLE
Move all the plugin object arguments to PluginContainer

### PR DESCRIPTION
**Pull request by ImMorpheus**
@Cybermaxke You raised the same point a while back. Did I miss something ?

Fix https://github.com/SpongePowered/SpongeAPI/issues/1721
Close https://github.com/SpongePowered/SpongeAPI/issues/892

### CODE REVIEW

**Comment by Lignium** [comment]
You have already fixed some of them, but I anyway made a complete list.

Search by "**plugin**" keyword:

```
AssetManager#getAsset(Object plugin, String name)
ChannelRegistrar#createChannel(Object plugin, String channel)
ChannelRegistrar#createRawChannel(Object plugin, String channel)
ChannelRegistrar#getOrCreate(Object plugin, String channel)
ChannelRegistrar#getOrCreateRaw(Object plugin, String channel)
ChunkPreGenerate.Builder#owner(Object plugin)
ChunkTicketManager#createEntityTicket(Object plugin, World world)
ChunkTicketManager#createPlayerEntityTicket(Object plugin, World world, UUID player)
ChunkTicketManager#createPlayerTicket(Object plugin, World world, UUID player)
ChunkTicketManager#createTicket(Object plugin, World world)
ChunkTicketManager#getAvailableTickets(Object plugin, World world)
ChunkTicketManager#getMaxTickets(Object plugin)
ChunkTicketManager#registerCallback(Object plugin, ChunkTicketManager.Callback callback)
CommandManager#register(Object plugin, CommandCallable callable, List<String> aliases)
CommandManager#register(Object plugin, CommandCallable callable, List<String> aliases, Function<List<String>,List<String>> callback)
CommandManager#register(Object plugin, CommandCallable callable, String... alias)
EventManager#registerListener(Object plugin, Class<T> eventClass, EventListener<? super T> listener)
EventManager#registerListener(Object plugin, Class<T> eventClass, Order order, boolean beforeModifications, EventListener<? super T> listener)
EventManager#registerListener(Object plugin, Class<T> eventClass, Order order, EventListener<? super T> listener)
EventManager#registerListener(Object plugin, TypeToken<T> eventType, EventListener<? super T> listener)
EventManager#registerListener(Object plugin, TypeToken<T> eventType, Order order, boolean beforeModifications, EventListener<? super T> listener)
EventManager#registerListener(Object plugin, TypeToken<T> eventType, Order order, EventListener<? super T> listener)
EventManager#registerListeners(Object plugin, Object obj)
EventManager#unregisterPluginListeners(Object plugin)
GameRegistry#registerAITaskType(Object plugin, String id, String name, Class<? extends AbstractAITask<? extends Agent>> aiClass)
Inventory.Builder#build(Object plugin)
MetricsConfigManager#areMetricsEnabled(Object plugin)
PermissionService#newDescriptionBuilder(Object plugin)
Scheduler#createAsyncExecutor(Object plugin)
Scheduler#createSyncExecutor(Object plugin)
Scheduler#getScheduledTasks(Object plugin)
ServiceManager#setProvider(Object plugin, Class<T> service, T provider)
ShapedCraftingRecipe.Builder.EndStep#build(String id, Object plugin)
ShapelessCraftingRecipe.Builder.EndStep#build(String id, Object plugin)
SimpleServiceManager#setProvider(Object plugin, Class<T> service, T provider)
SqlService#getDataSource(Object plugin, String jdbcConnection)
Task.Builder#submit(Object plugin)
TimingsFactory#of(Object plugin, String name, Timing groupHandler)
Timings#of(Object plugin, String name)
Timings#of(Object plugin, String name, Timing groupHandler)
Timings#ofStart(Object plugin, String name)
Timings#ofStart(Object plugin, String name, Timing groupHandler)
```

Search by "**instance**" keyword:

```
CommandManager#getOwnedBy(Object instance)
ConfigManager#getPluginConfig(Object instance)
ConfigManager#getSharedConfig(Object instance)
PluginManager#fromInstance(Object instance)
```

**Comment by ImMorpheus** [comment]
@Lignium 

Just a quick note:

This is an API PR, as such I'm focusing on the API.
You need to be careful, some of those methods are used by mods too (see `ChunkTicketManager`), I can't blindly replace everything.

Thank you for the list. I've missed a few, I will fix it soon.

**Comment by Lignium** [comment]
Sponge API is used by only plugins, isn't it? Mods can only use Forge API.

**Comment by dualspiral** [comment]
> Sponge API is used by only plugins, isn't it? Mods can only use Forge API.

No, mods can freely use the SpongeAPI if they so wish. Note that they can [use their `ModContainer` as the `PluginContainer` as we mixin the `PluginContainer` interface into `ModContainer`.](https://github.com/SpongePowered/SpongeForge/blob/stable-7/src/main/java/org/spongepowered/mod/mixin/core/fml/common/MixinModContainer.java#L50-L51)

**Comment by ImMorpheus** [comment]
Sorry, I worded it poorly. 
Let me rephrase that: part of the API is implemented in SpongeForge and it's using Forge methods/classes.

https://github.com/SpongePowered/SpongeAPI/blob/6632f971b9a3b04aa839ed28ce4839e02dc32b9b/src/main/java/org/spongepowered/api/world/ChunkTicketManager.java#L57

https://github.com/SpongePowered/SpongeForge/blob/1544c736731f603d7818e5aff1db7090a2085fd9/src/main/java/org/spongepowered/mod/service/world/SpongeChunkTicketManager.java#L62

In this case, I can't change the signature to PluginContainer. 

EDIT: dual explained the part about mods being able to use the SpongeAPI

**Comment by Lignium** [comment]
@ImMorpheus But you may well change the implementation of this method, by using a call of `PluginContainer#getInstance()` to convert from container to instance.

**Comment by ImMorpheus** [comment]
There would be no point in changing the argument type to PluginContainer then. Why convert back and forth between PluginContainer and Object ?

EDIT: nope, dual raised a good point

**Comment by dualspiral** [comment]
I'm inclined to agree. The fact that Forge needs the actual plugin instance is an implementation detail really and as @Lignium said, we should make the API accept a `PluginContainer` if we are doing that everywhere else and deal with the conversion in SC/SF.

**Comment by Cybermaxke** [comment]
@ImMorpheus The issue https://github.com/SpongePowered/SpongeAPI/issues/1721 was about the fact that we were requiring the plugin instance/id and id in the `build` method and doing this inconsistently.
That issue got fixed in https://github.com/SpongePowered/SpongeAPI/commit/d40b52e9a046a02d072af660b52f589d46f87e0c, I just forgot to close it

**Comment by Lignium** [comment]
I hope you will not reject this pull request?

**Comment by Cybermaxke** [comment]
I will not close it, but it's unrelated.

These are good changes because you can no longer pass in just any object. But it could be annoying that you always have to pass in the container.

**Comment by FerusGrim** [comment]
Previously, the majority of these methods required the actual `@Plugin` annotated instance to be passed to them, not the `PluginContainer`, unless I'm mistaken. In that vein, shouldn't the javadocs reflect that we're not passing the plugin instance but, rather, the plugin container? The container is now being referred to as the plugin instance and while it should be obvious given the method signature what's required, it seems inconsistent.

Feel free to ignore my ramblings if I'm wildly off-base.

**Comment by Cybermaxke** [comment]
The current `Object` can be the `PluginContainer` or the `@Plugin` annotated instance. Both are valid options.
